Import ISLET project for Arm CCA

bitboom opened the pull request:
First, I am glad to collaborate with `Certifier` :)

This PR imports `ISLET SDK` and adds the sample application.
This version runs on x86_64 to integrate easily.

My plan is like:
1. Import ISLET SDK and make the sample app (simulated version on x86_64) << **This PR**
2. Integrate interfaces between `ceritifer` and `islet` (x86_64) << *I hope that certifier developer could help this!*
3. Run sample app on Arm CCA simulator (arm64)

## ISLET SDK
It is for applications running on Arm CCA. ISLET SDK is written in `rust` but we also support `C/C++`.
We use `cbindgen` which is the powerful tool to create headers for rust libraries which expose public C/C++ APIs.
_So please do not modify header directly._

Please refer [this](https://github.com/Samsung/islet/tree/main/sdk) for details.

```
+--------+                           +---------+
| sdk    | => cbindgen => header  => | app     |
| (rust) | => cdylib   => library => | (c/c++) |
+--------+                           +---------+
```

Feel free to give me feedback and any questions.

bitboom commented:
I am trying to find proper the open-source manager in my company. After that, I am going to sign the `contributor license agreement`.

gapisback commented on sample_apps/islet_attest_seal/README.md:
```diff
@@ -0,0 +1,21 @@
+# A sample app towards Arm CCA
+This sample app is towards Arm CCA with [ISLET SDK](https://github.com/Samsung/islet/tree/main/sdk)
+Currently this is running on x86_64 to intergrate easily. 
```
Typo: 'integrate'

gapisback commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+    else
+        printf("[!] Failed to attestation test...\n");
+
+    printf("[!] Starting sealing test...\n");
```
Qs here: If `attestation_test()` were to FAIL, why would we want to continue wiht sealing test? 

Do we not want to exit if `attestation_test()`  returns fals?

bitboom replied:
The reason is that those are orthogonal test. But i think it is better to exit when `attestation_test()` is failed.

gapisback commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+    if (sealing_test()) 
+        printf("[!] Success to sealing test...\n");
+    else
+        printf("[!] Failed to sealing test...\n");
```
Please rewrite this block as suggested previously, using `bool rv` and ternary for `printf()`.

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved.
```
Oh, I see this copyright here. Do you [not] need to mention the license under which this is copyrighted?
Similar copyright notice may be needed in `.cc`, `.c` file(s).

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,69 @@
+
+constexpr static const int ISLET_E_WRONG_CLAIMS = -4;
+
+constexpr static const int ISLET_E_NOT_SUPPORTED = -5;
```
Is it possible to rewrite these constants as enum-values?

Say, you define a typedef `islet_rv_t`, and define all these as this enumerated type? Not sure if this is feasible in this C++ code.

Enum-types are helpful as compiler will help checking for returned value. Also, you can see the enum-type-name displayed while debugging.

bitboom replied:
Thanks for suggestion! I added `islet_status_t` enum :)

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,69 @@
+/// # Note
+/// This API returns hard coded report to simuate temporarily.
+/// `User data` could be used as nonce to prevent reply attack.
+int islet_attest(const unsigned char *user_data,
```
If you do choose the suggestion of enum-types, I think all these functions' definition should change to return `islet_rv_t`.

bitboom replied:
I fixed it :)

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,69 @@
+/// Seals the plaintext given into the binary slice
+///
+/// # Note
+/// This API seals with hard coded key to simuate temporarily.
```
Typo: 'simulate'.

I think a rephrased message would be better:

```
This API currently seals with a hard-coded key, to simulate seal operation. 
In future, this will be finalized to support _<fill-in-the-blank>_ keys.
```

Replace the _<fill-in-the-blank>_  with some term better describing what you intend to do in the future.

Hope this rewrite makes sense to you ...

bitboom replied:
Thanks for suggestion! I reflected it.

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,69 @@
+               unsigned char *sealed_out,
+               int *sealed_out_len);
+
+/// Unseals the sealed bianry given into the plaintext
```
Typo: 'binary'. 

Maybe rephrase as: `Unseals into plaintext the sealed binary provided.`

bitboom replied:
I reflected it! Thanks.

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,69 @@
+/// Unseals the sealed bianry given into the plaintext
+///
+/// # Note
+/// This API unseals with hard coded key to simuate temporarily.
```
Typo: 'simulate'.

You may want to rephrase this comment, also, based on the suggestion given on L54 for the seals API comment.

bitboom replied:
I fixed it. Thanks!

gapisback commented on third_party/islet/setup.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
```
Please add a small banner prolog comment documenting what this shell script is for / about.

You may want to follow the convention introduced (by me) recently in `sample_apps/run_example.sh`.

gapisback commented on third_party/islet/setup.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+set -e
```
This is good, but over time, ... I have been taught (& discovered) that a more stricter checking for errors during shell script execution is better.

See if you can expand this to `set -Eeuo pipefail`, so that any errors in any command will cause this script to abort immediately.

This business of `$rc` error checking in bash scripts is bit difficult to work with. I experimented with some of this most recently in above-mentioned shell script. You may want to review what I did, to see if that model will work for your script. And adapt that accordingly.

gapisback replied:
Also it is a good idea to always run `shellcheck` on this script, and fix all warnings / errors reported. Have you done that?

bitboom replied:
I didn't use `shellcheck`. (I actually have used `shfmt` on my projects).
Thanks for valuable comments! I reflected both comments.

gapisback commented on third_party/islet/setup.sh:
```diff
@@ -0,0 +1,30 @@
+
+set -e
+
+ROOT=$(git rev-parse --show-toplevel)
```
Is this better named as `CC_ROOT`, for clarity that this is the top-level git-sub-dir for CC library?

(Or, if this is some other repo, consider adding a prefix to indicate what this `root` dir is for.)

bitboom replied:
I changed it to `CC_ROOT`.

gapisback commented on third_party/islet/setup.sh:
```diff
@@ -0,0 +1,30 @@
+
+mkdir -p $ISLET_INC $ISLET_LIB
+cp $TARGET_HDR $ISLET_INC
+cp $TARGET_LIB $ISLET_LIB
```
Please try to always use `cp -p` so that the file's timestamp is preserved. (Helps script debugging.)

bitboom replied:
I attached `-p` option.

gapisback reviewed (commented):
Hi, @bitboom -- I am a developer on the CC library with VMware.

I wanted to discover and learn what your team has done with this Islet integration.

So, I took the opportunity to review this code based on our discussion in the team meeting last night, with @jlmucb .

I left my review as 'open', expecting that you will study the comments and adjust accordingly. Once John gets to see this, feel free to discuss with him how far to go toward implementing the suggestions given. (Most of my changes are stylistic, and clean-up type, so don't expect that they will take too much effort to update the code.)

Thanks.
--AdityA>

gapisback commented:
Attn: @bitboom -- I see you have 3 commits wanting to merge to `/main`. Once the code reviews are done, please do remember to squash all commits into one, and add a single commit-message.

In the UI, you should be able to pick "Squash and rebase".

Or, you can do squash offline in your dev-branch, and update the commit-message to describe briefly the changes done.
Then force-push this back to your dev-branch and pick "Rebase and merge" (or some such button) in this UI to push this to `/main`.

jlmucb commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+    memset(report, 0, sizeof(report));
+    memset(claims, 0, sizeof(report));
+    memset(value, 0, sizeof(value));
+    int report_len = 0, claims_len = 0, value_len = 0;
```
yes.   They c++ style is generally the "Google C++" style which Google has published

jlmucb replied:
Incidentially, in the Google C++ style, the "tab" is 2 spaces, not 4.

bitboom replied:
I formatted it by `astyle --style=google --indent=spaces=2`. Thanks for letting me know Coding Style what certifier followed :+1:

jlmucb commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+#include <islet.h>
```
The copyright notice for Samsung contributions should be (c) Samsung and say it is licensed under the Apache license (or if another license, it shoudl say and include the license in thedirectory).

bitboom commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+static const char CLAIM_TITLE_PLATFORM_PROFILE[] = "Profile";
+
+bool attestation_test() {
+    byte report[2048], claims[1024], value[1024];
```
I defined `BUFFER_SIZE` :)

bitboom commented on sample_apps/islet_attest_seal/README.md:
```diff
@@ -0,0 +1,21 @@
+
+## Getting started
+### Setup islet sdk
+ISLET SDK is written in rust but it also support C/C++.
```
Thanks for suggestion! I fixed it!

bitboom commented on sample_apps/islet_attest_seal/README.md:
```diff
@@ -0,0 +1,21 @@
+# A sample app towards Arm CCA
+This sample app is towards Arm CCA with [ISLET SDK](https://github.com/Samsung/islet/tree/main/sdk)
+Currently this is running on x86_64 to intergrate easily. 
+The real app running on Arm CCA simulator (arm64) is comming soon.
```
I fixed it.

bitboom commented on sample_apps/islet_attest_seal/README.md:
```diff
@@ -0,0 +1,21 @@
+# A sample app towards Arm CCA
+This sample app is towards Arm CCA with [ISLET SDK](https://github.com/Samsung/islet/tree/main/sdk)
+Currently this is running on x86_64 to intergrate easily. 
```
Thanks!

bitboom commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+}
+
+bool sealing_test() {
+    byte sealed[2048], unsealed[2048];
```
I patched it. Thanks!

bitboom commented on sample_apps/islet_attest_seal/attest_seal_app.cc:
```diff
@@ -0,0 +1,79 @@
+    if (attestation_test()) 
+        printf("[!] Success to attestation test...\n");
+    else
+        printf("[!] Failed to attestation test...\n");
```
It looks more concise. Thanks!

bitboom commented:
> Attn: @bitboom -- I see you have 3 commits wanting to merge to `/main`. Once the code reviews are done, please do remember to squash all commits into one, and add a single commit-message.
> 
> In the UI, you should be able to pick "Squash and rebase".

@gapisback 
I will squash commits into one by using UI as you mentioned.
And i reflected your comments. Really thank you for kind comments!

bitboom commented:
@jlmucb 
I received the email and thank you so much for the warm and friendly welcome 😄.
I replied to your comments below. 

Please let me know if there are other comments.

---

> There are a bunch of comments we left in the github repository.  The main comment is that this will turn into interface code and interface code test, not a "sample app,"  so it should go in a new directory $CERTIFIER/src/cca, for example.  Later, there will be a directory in sample_apps called simple_app_under_cca and maybe other sample apps you contribute.

I moved interface code test moved to `src/cca/islet_test`.

> A minor comment:  The C++ style we use is the Google style (we use Google Go style too).  It's easy to find on the web.  2 spaces for tab, usually single variable declaration per line although we're not dictatorial.  

I reflected style with `astyle --style=google --indent=spaces=2 *.cc` . And I also fixed it to `single variable declaration`.

> The files you contribute in their entirety should bear a Samsung copyright notice and say they are licensed under the Apache license in the main directory.  If you want to license under BSD, you can do that but we should include a copy of that license in the main directory, fo reference in that case.

I added a Samsung Copyright notice under the Apache license  to `*.h,  *.cc`.

bitboom commented:
@jlmucb I have finally done CLA process. :smile: 
Could you please check this PR can be merged?

---

FYI, I recently done to get `real attestation report` from `CCA simulator` at https://github.com/Samsung/islet/pull/76.
So, i will make next PR to certifier which is run `islet test` on CCA simulator.

gapisback commented on src/cca/islet_test/attest_seal_test.cc:
```diff
@@ -0,0 +1,102 @@
+#include <string>
+#include <cstring>
+
+#define BUFFER_SIZE 2048
```
Please, can you document where we arrived at this limit from? The reason for these kinds of sizes will be lost over time, so even if you can add a simple comment, something like:

// Some reasonable size to allocate <... fill-in-the-blank> ... on-stack buffers.

If you can give the reason why 2K is big enough, that will be helpful. (For future readers.)

bitboom replied:
I added it. Thank you for the suggestion!

gapisback commented on src/cca/islet_test/attest_seal_test.cc:
```diff
@@ -0,0 +1,102 @@
+  int value_len = 0;
+
+  memset(report, 0, sizeof(report));
+  memset(claims, 0, sizeof(report));
```
Although this will work, technically this should be : `sizeof(claims)`. Please fix.

bitboom replied:
You're right. I fixed it.

gapisback commented on third_party/islet/include/islet.h:
```diff
@@ -0,0 +1,83 @@
+islet_status_t islet_attest(const unsigned char *user_data,
+                            int user_data_len,
+                            unsigned char *report_out,
+                            int *report_out_len);
```
Thank you for the nice indentation of fn-arguments. It makes the `.h` very much more readable.

gapisback commented on third_party/islet/setup.sh:
```diff
@@ -0,0 +1,34 @@
+# setup.sh - Setup script to build ISLET SDK
+# ##############################################################################
+
+set -Eeuo pipefail
```
(FYI: I'm assuming you've run `shellcheck` on this script. Right now, there are only a few Bash scripts. I am thinking it will be good to invoke `shellcheck` as part of CI, so that every new script change will adhere to common formatting. 

That's not in-place, just yet ... but will be good if have run this thru `shellcheck` while we are adding a new script.
)

bitboom replied:
Yes. It's very good to add `shellcheck` to CI. 
And  c++ conding style check on CI is also good too =)

gapisback reviewed (approved):
Took a quick re-look at your updated files. I've put down few very minor comments, which, if you can address, it will be good.

Thanks for improving the code based on my previous set of minor comments. Things are beginning to look more polished with this rev.

bitboom commented on third_party/islet/setup.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+# ##############################################################################
+# setup.sh - Setup script to build ISLET SDK
+# ##############################################################################
```
Thank you for pointing out. I added the copyright.

bitboom commented:
> Can you remove the build.yml changes (but save the file) from the pull request? Were tidying up the yml tests. If you do that, you can merge (or I'll merge for you). Everything else is fine.
> 
> Thanks

I removed it =)
